[c10d] Make barrier as a custom op

alanwaketan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #79779
* __->__ #79777
* #79691

Summary:
This patch makes barrier as a custom op such that it's dispatcher
passable. It's one part of the effort to route comm ops to the dispatcher
such that tracing mechanisms that relies on the dispatcher can trace them,
e.g., LazyTensor and AOTAutograd.

Test Plan:
python test/distributed/test_c10d_nccl.py -k test_nccl_barrier
...and other existing distributed tests.

Commit:
[c10d] Make alltoall as a custom op

Summary:
This patch makes alltoall as a custom op such that it's dispatcher
passable. It's one part of the effort to route comm ops to the dispatcher
such that tracing mechanisms that relies on the dispatcher can trace them,
e.g., LazyTensor and AOTAutograd.

Test Plan:
python test/distributed/test_c10d_nccl.py -k test_nccl_barrier
...and other existing distributed tests.

[ghstack-poisoned]

Commit:
Update on "[c10d] Make barrier as a custom op"


Summary:
This patch makes barrier as a custom op such that it's dispatcher
passable. It's one part of the effort to route comm ops to the dispatcher
such that tracing mechanisms that relies on the dispatcher can trace them,
e.g., LazyTensor and AOTAutograd.

Test Plan:
python test/distributed/test_c10d_nccl.py -k test_nccl_barrier
...and other existing distributed tests.

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/c10d/Ops.cpp:
```diff
@@ -142,6 +150,9 @@ TORCH_LIBRARY(c10d, m) {
       "alltoall_",
       dispatch(c10::DispatchKey::CompositeExplicitAutograd, alltoall_));
+  m.def(
+      "barrier",
```
missing a trailing underscore?

alanwaketan replied:
Ditto.

alanwaketan commented on torch/csrc/distributed/c10d/Ops.cpp:
```diff
@@ -110,6 +110,14 @@ c10::intrusive_ptr<ProcessGroup::Work> alltoall_(
       AllToAllOptions{std::chrono::milliseconds(timeout)});
 }
 
+c10::intrusive_ptr<ProcessGroup::Work> barrier(
```
barrier is not a in-place op. Therefore, it's not necessary to add a trailing _.

mrshenli replied:
Oh I see, trailing score is for inplace ops as in Python APIs. Thought it was a dispatcher convention.

Commit:
Update on "[c10d] Make barrier as a custom op"


Summary:
This patch makes barrier as a custom op such that it's dispatcher
passable. It's one part of the effort to route comm ops to the dispatcher
such that tracing mechanisms that relies on the dispatcher can trace them,
e.g., LazyTensor and AOTAutograd.

Test Plan:
python test/distributed/test_c10d_nccl.py -k test_nccl_barrier
...and other existing distributed tests.

[ghstack-poisoned]

alanwaketan commented:
@pytorchbot merge --green
